api/src/main/java/marquez/db/LineageDao.java

```diff
@@ -154,4 +171,41 @@ SELECT DISTINCT on(r.job_name, r.namespace_name) r.*, jv.version as job_version
       WHERE j.uuid in (<jobUuid>) OR j.symlink_target_uuid IN (<jobUuid>)
       ORDER BY r.job_name, r.namespace_name, created_at DESC""")
   List<Run> getCurrentRuns(@BindList Collection<UUID> jobUuid);
+
+  @SqlQuery(
+      """
+WITH RECURSIVE
+  upstream_runs(
+          r_uuid, started_at, ended_at, state,
+          job_uuid, job_version_uuid, job_namespace, job_name,
+          dataset_uuid, dataset_version_uuid, dataset_namespace, dataset_name,
+          u_r_uuid, depth) AS (
+    SELECT
+          r.uuid, r.started_at, r.ended_at, r.current_run_state,
+          r.job_uuid, r.job_version_uuid, r.namespace_name, r.job_name,
+          dv.dataset_uuid, dv."version", dv.namespace_name, dv.dataset_name,
+          dv.run_uuid,
+          0 AS depth
+    FROM runs r
+    LEFT JOIN runs_input_mapping rim ON rim.run_uuid = r.uuid
+    LEFT JOIN dataset_versions dv ON dv.uuid = rim.dataset_version_uuid
```

Comment:
Just thinking loudly. Wouldn't be better to join dataset_versions after the recursion at the bottom of the query once all the runs are identified?

Reply:
I thought of that as well but we actually need dataset_versions in the recursion because this is where we find the run_uuid that produced the DV for the next iteration.

Reply:
idea: we could add that run uuid to runs_input_mapping at the same time as the dataset_version that would allow to join just on that table in the recursion. That'd be neat.

Reply:
Didn't get the idea: do you want to add dataset_versions to runs_input_mapping or run_uuid to dataset_versions?
Anyway, the current state of the query looks good to me.

Reply:
The idea is we we could add the dataset_versions.run_uuid to the runs_input_mapping table. 
The run_uuid is always already set in the dataset_versions table when we write the dataset_versions.uuid to the runs_input_mapping table and it never changes since it is the run that created that dataset version.
Then you just need to recursively join the runs_input_mapping over itself

Reply:
but I would keep that for future improvement.